samples: gettimeofday: Increase stack size by 64 bytes and other improvements

pfalcon opened the pull request:
Otherwise the sample started to fail on qemu_x86.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
samples: gettimeofday: Increase stack size by 64 bytes.

Otherwise the sample started to fail on qemu_x86.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

andrewboie commented:
This sample application is missing a sample.yaml, it is never being built by CI for any platform. Can you please add this so it doesn't continue to bit-rot?

pabigot commented:
It also produces build diagnostics due to `return 1;` in `void main(void)`.

The change looks plausible, but I can't [make it run](https://docs.zephyrproject.org/latest/samples/posix/gettimeofday/README.html) to verify the problem and solution:
```
tirzah[10]$ west build -t run
-- west build: running target run
[0/1] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: qemu32,+nx,+pae
qemu-system-i386: -serial unix:/tmp/slip.sock: Failed to connect socket /tmp/slip.sock: No such file or directory
qemu-system-i386: -serial unix:/tmp/slip.sock: could not connect serial device to character backend 'unix:/tmp/slip.sock'
FAILED: zephyr/CMakeFiles/run 
cd /mnt/nordic/zp/zephyr/build && /usr/local/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/bin/qemu-system-i386 -m 9 -cpu qemu32,+nx,+pae -device isa-debug-exit,iobase=0xf4,iosize=0x04 -no-reboot -nographic -net none -pidfile qemu.pid -chardev stdio,id=con,mux=on -serial chardev:con -mon chardev=con,mode=readline -serial unix:/tmp/slip.sock -kernel /mnt/nordic/zp/zephyr/build/zephyr/zephyr.elf
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/bin/cmake --build /mnt/nordic/zp/zephyr/build --target run
```

pfalcon commented:
> This sample application is missing a sample.yaml, it is never being built by CI for any platform. Can you please add this so it doesn't continue to bit-rot?

That wouldn't help with the current issue (as it's runtime one), but yep, working to add sample.yaml.

pfalcon commented:
> It also produces build diagnostics due to return 1; in void main(void).

That's conscious regression introduced in https://github.com/zephyrproject-rtos/zephyr/pull/21100, which was merged regardless of all the concerns raised.

pfalcon commented:
> The change looks plausible, but I can't make it run to verify the problem and solution:

As someone experienced with Zephyr, I'm sure it didn't escape you that it needs networking setup, but yeah, that should be in the README, working on it.

pfalcon commented:
And it doesn't end there:

~~~
[ 43%] Building C object zephyr/CMakeFiles/zephyr.dir/subsys/net/lib/sntp/sntp.c.obj
In file included from /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/subsys/net/lib/sntp/sntp.c:12:
/home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/subsys/net/lib/sntp/sntp_pkt.h:14: error: "MODE_MASK" redefined [-Werror]
   14 | #define MODE_MASK 0x07
      | 
In file included from /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/include/arch/arm/aarch32/asm_inline_gcc.h:26,
                 from /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/include/arch/arm/aarch32/asm_inline.h:18,
                 from /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/include/arch/arm/aarch32/arch.h:33,
                 from /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/include/arch/cpu.h:19,
                 from /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/include/kernel_includes.h:39,
                 from /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/include/kernel.h:17,
                 from /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/include/net/net_context.h:23,
                 from /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/include/net/dns_resolve.h:17,
                 from /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/include/net/socket.h:27,
                 from /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/include/posix/sys/socket.h:10,
                 from /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/include/net/sntp.h:14,
                 from /home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/subsys/net/lib/sntp/sntp.c:11:
/home/pfalcon/projects-3rdparty/Embedded/Zephyr/zephyr/include/arch/arm/aarch32/cortex_a_r/cpu.h:17: note: this is the location of the previous definition
   17 | #define MODE_MASK 0x1f
      | 
~~~

So, Zephyr in its full glory, where touching one issue leads to 3 levels deep of them.

What's interesting about this last one is that it's 3rd namespacing issue I see around during last few weeks. (But of course, no conclusions will be drawn, just the same as none were drawn when it was raised (speculatively) years ago.)

pfalcon committed:
samples: posix: gettimeofday: Add sample.yaml

To make sure this sample is built in CI.

Also includes fixing prototype of main() to make it build without
warnings (which would be errors in CI).

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
samples: posix: gettimeofday: Make networking requirements explicit

Standard "Requirements" section copied from other networking samples,
making clear the network connection requirements and linking to
instructions for QEMU networking.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
This now should contain fixes for all issues mentioned, including #24756, which is being bikeshedded so far.

pabigot commented on samples/posix/gettimeofday/README.rst:
```diff
@@ -12,6 +12,14 @@ system startup, the current time is queried using the SNTP networking
 protocol, enabled by setting the :option:`CONFIG_NET_CONFIG_CLOCK_SNTP_INIT`
 and :option:`CONFIG_NET_CONFIG_SNTP_INIT_SERVER` options.
 
+Requirements
```
tl;dr: this is a clue, but not adequate for newbies.  Since it's what's done for other samples it's at least meeting the standard of practice.

I looked at "Networking with host", selected [QEMU Ethernet] https://docs.zephyrproject.org/latest/guides/networking/qemu_eth_setup.html#networking-with-eth-qemu)
because it seemed simplest, and verified `samples/net/sockets/echo_server` worked using telnet to 192.0.2.1:4242
(oddly the [sample recommended use of `echo-client`](https://docs.zephyrproject.org/latest/samples/net/sockets/echo_server/README.html#running-echo-client-in-linux-host) produced a `SO_BINDTODEVICE: No such device`, so that sample isn't completely documented either).

The gettimeofday sample, of course, doesn't use that variant of host networking; it seems to *require* the SLIP interface ("legacy...use only when necessary").

Following that procedure almost worked, except that I'm not running an SNTP server on my development host so it couldn't set the time:

```
gettimeofday(): HI(tv_sec)=0, LO(tv_sec)=5, tv_usec=30000
[00:00:05.030,000] <err> net_config: Cannot set time using SNTP
gettimeofday(): HI(tv_sec)=0, LO(tv_sec)=6, tv_usec=40000
```

For completeness and to assist people who do not normally use networking this should specify that "QEMU using SLIP" should be used, and that the local host must run an SNTP server.

(Unless the Ethernet networking is supposed to work, in which case that's a different problem.)

pabigot reviewed (approved):
I've expressed unhappiness with documentation but I don't expect that to be addressed; it's in range for existing POSIX/net sample documentation.  Functionally this looks fine.  I confirmed that the `sample.yaml` does cause this to be built in CI, even though it doesn't with `sanitycheck -t samples/posix/gettimeofday`.

andrewboie commented:
>  I confirmed that the `sample.yaml` does cause this to be built in CI, even though it doesn't with `sanitycheck -t samples/posix/gettimeofday`.

@pabigot `-t` is `--tag`. You want `-T` which is `--testcase-root`

jukkar commented on subsys/net/lib/sntp/sntp_pkt.h:
```diff
@@ -9,21 +9,21 @@
+#define Z_SNTP_LI_MASK   0xC0
+#define Z_SNTP_VN_MASK   0x38
+#define Z_SNTP_MODE_MASK 0x07
 
```
Could you remove the SNTP naming changes from this PR?

pfalcon replied:
> Could you remove the SNTP naming changes from this PR?

Can you please approve https://github.com/zephyrproject-rtos/zephyr/pull/24756, and ideally merge it, then I'll rebase this PR on master.

pfalcon commented:
Rebased on the latest master. Should be ready to merge.
